Wikipedia: save full page as epub (html, with optional images)

poire-z opened the pull request:
... and TOC, and cover image. API and html from Wikipedia should be stable, so this should hopefully not need much maintenance. There are only a few tweaks to the HTML to work around some small crengine quirks.
Sample generated epub available in #2493

Included a zipwriter.lua that i put in frontend/ (did not find an existing other suitable place). It is based on some other works (but these can't be used as is), mentionned at top of file. Not sure that is enough to give credits, or if other things are needed.

Also for the current text-only wikipedia full page: replace the ==s in section titles with some unicode symbols for a better visual feeling of hierarchy. These same symbols are also used in the .epub.
Both can be disabled by adding ["wikipedia_prettify"] = false to settings.reader.lua
You'll tell me if that should not be the default.

readerhighlight: close ButtonTable (like other buttons do) when looking up wikipedia, otherwise if we save as epub, and swtich to new document, this ButtonTable will never be closed and stays in UI.

Frenzie commented on frontend/ui/widget/dictquicklookup.lua:
```diff
@@ -245,11 +245,68 @@ function DictQuickLookup:update()
+                        local lang = self.lang or self.wiki_languages_copy[1]
+                        -- Just to be safe (none of these chars, except ':' for uninteresting
+                        -- Portal: or File: wikipedia pages, should be in lookup_word)
+                        local cleaned_lookupword = self.lookupword:gsub('[\\,%/,:,%*,%?,%",%<,%>,%|]','_')
```
Didn't we add a utility function just for this?

Frenzie replied:
`util.replaceInvalidChars`

https://github.com/koreader/koreader/commit/7bfbc9f06a4a1485c0c9280a365197430477a9ee

poire-z replied:
I think i may have cut and pasted it from there :) Guess I liked having that gsub there in case i needed more. But ok, I will use util.replaceInvalidChars at next rebase.

Frenzie replied:
If more were needed (which it isn't :-P) then surely it would indicate a bug that should be fixed in util.replaceInvalidChars? :-)

houqp commented:
luarocks' license is GPL compatible, so it should be fine.

I like how little code it's needed to generate the epub :) Generally, I would write the wiki format "transplier" with lpeg instead of regex substitution for better performance, memory usage and maintainability. This can be left for future cleanup.

poire-z commented:
Done: bumbed base and uses `ffi/zipwriter.lua`, uses `util.replaceInvalidChars()`, removed commented `logger.dbg()`, and more uses `of string.format()`.
Also moved the removal of `<a>` around `<img>` from the section we deal with images to later on when we fix HTML, as I just saw there could be images not wrapped in a `<a>` that were missed.
Also added gsub() to add `https://` to external links with `href="//en.wikipedia.org"` .
